### PR TITLE
Add `iam_groups` to `vault_aws_secret_backend_role`

### PR DESCRIPTION
**Pull request by lawliet89**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-vault/blob/master/CHANGELOG.md):

```release-note
Add `iam_groups` to `vault_aws_secret_backend_role`
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSecretBackendRole'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v -run=TestAccAWSSecretBackendRole -timeout 120m
?       github.com/terraform-providers/terraform-provider-vault [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/coverage    [no test files]
?       github.com/terraform-providers/terraform-provider-vault/cmd/generate    [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/codegen (cached) [no tests to run]
?       github.com/terraform-providers/terraform-provider-vault/generated       [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/datasources/transform/decode  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/datasources/transform/encode  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/alphabet  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/role      (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/template  (cached) [no tests to run]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/generated/resources/transform/transformation    (cached) [no tests to run]
?       github.com/terraform-providers/terraform-provider-vault/schema  [no test files]
testing: warning: no tests to run
PASS
ok      github.com/terraform-providers/terraform-provider-vault/util    (cached) [no tests to run]
=== RUN   TestAccAWSSecretBackendRole_basic
--- PASS: TestAccAWSSecretBackendRole_basic (0.41s)
=== RUN   TestAccAWSSecretBackendRole_import
--- PASS: TestAccAWSSecretBackendRole_import (0.32s)
=== RUN   TestAccAWSSecretBackendRole_nested
--- PASS: TestAccAWSSecretBackendRole_nested (0.33s)
PASS
ok      github.com/terraform-providers/terraform-provider-vault/vault   (cached)

...
```

### CODE REVIEW

File: vault/resource_aws_secret_backend_role_test.go [hunk 1]
```diff
@@ -243,12 +251,14 @@ resource "vault_aws_secret_backend_role" "test_policy_inline" {
...
   default_sts_ttl = 3600
   max_sts_ttl = 21600
+  iam_groups = ["group1", "group2"]
 }
 
 resource "vault_aws_secret_backend_role" "test_policy_arns" {
   name = "%s-policy-arn"
   policy_arns = ["%s"]
-  credential_type = "iam_user"
+	credential_type = "iam_user"
```
Comment (tvoran) [inline vault/resource_aws_secret_backend_role_test.go:260]:
```suggestion
  credential_type = "iam_user"
```

File: vault/resource_aws_secret_backend_role_test.go [hunk 2]
```diff
@@ -257,13 +267,15 @@ resource "vault_aws_secret_backend_role" "test_policy_inline_and_arns" {
   policy_document = %q
   policy_arns = ["%s"]
   credential_type = "iam_user"
-  backend = "${vault_aws_secret_backend.test.path}"
+  iam_groups = ["group1", "group2"]
+	backend = "${vault_aws_secret_backend.test.path}"
```
Comment (tvoran) [inline vault/resource_aws_secret_backend_role_test.go:271]:
```suggestion
  backend = "${vault_aws_secret_backend.test.path}"
```

File: vault/resource_aws_secret_backend_role_test.go [hunk 3]
```diff
@@ -257,13 +267,15 @@ resource "vault_aws_secret_backend_role" "test_policy_inline_and_arns" {
...
 }
 
 resource "vault_aws_secret_backend_role" "test_role_arns" {
 	name = "%s-role-arns"
 	role_arns = ["%s"]
 	credential_type = "assumed_role"
+  iam_groups = ["group1", "group2"]
```
Comment (tvoran) [inline vault/resource_aws_secret_backend_role_test.go:278]:
```suggestion
	iam_groups = ["group1", "group2"]
```

**Comment by lawliet89** [comment]
Thanks. The spacing/tabs is hard to differentiate in my VSCode.

File: vault/resource_aws_secret_backend_role.go [hunk 4]
```diff
@@ -127,6 +135,11 @@ func awsSecretBackendRoleWrite(d *schema.ResourceData, meta interface{}) error {
 
 	credentialType := d.Get("credential_type").(string)
 
+	var iamGroups []interface{}
+	if v, ok := d.GetOk("iam_groups"); ok {
+		iamGroups = v.(*schema.Set).List()
+	}
```
Comment (tvoran) [inline vault/resource_aws_secret_backend_role.go:141]:
```suggestion
	iamGroups := d.Get("iam_groups").(*schema.Set).List()
```

File: vault/resource_aws_secret_backend_role.go [hunk 5]
```diff
@@ -139,6 +152,9 @@ func awsSecretBackendRoleWrite(d *schema.ResourceData, meta interface{}) error {
 	if len(roleARNs) != 0 {
 		data["role_arns"] = roleARNs
 	}
+	if len(iamGroups) != 0 {
```
Comment (tvoran) [inline vault/resource_aws_secret_backend_role.go:155]:
```suggestion
	if len(iamGroups) != 0 || !d.IsNewResource() {
```

**Comment by lawliet89** [comment]
Thanks for the suggestion.

I asked a member of the Vault team before and they said that if you send a parameter to a version of Vault that does not support it (in this case < 1.4),  Vault would ignore it. I think we might be able to simplify this by always sending the parameter whether the list is empty or not.

**Comment by catsby** [comment]
Hello all - this was released in [`v2.13.0`][v2130] , thanks for the contribution and your patience!


[v2130]: https://github.com/terraform-providers/terraform-provider-vault/releases/tag/v2.13.0

**Pull request merged**